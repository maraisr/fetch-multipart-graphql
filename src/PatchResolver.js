import { parseMultipartHttp } from './parseMultipartHttp';

export function PatchResolver({ onResponse }) {
    this.onResponse = onResponse;
    this.processedChunks = 0;
    this.chunkBuffer = '';
}

PatchResolver.prototype.handleChunk = function(data) {
    this.chunkBuffer += data;
    const { newBuffer, parts } = parseMultipartHttp(this.chunkBuffer);
    this.chunkBuffer = newBuffer;
    if (parts.length) {
        this.onResponse(parts);
    }
};
