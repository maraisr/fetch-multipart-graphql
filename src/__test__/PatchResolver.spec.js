import { PatchResolver } from '../PatchResolver';
import { TextEncoder, TextDecoder } from 'util';
// polyfill TextDecoder/Encoder for node
global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;

const chunk1 = [
    '',
    '---',
    'Content-Type: application/json',
    'Content-Length: 142',
    '',
    '{"data":{"viewer":{"currencies":null,"user":{"profile":null,"items":{"edges":[{"node":{"isFavorite":null}},{"node":{"isFavorite":null}}]}}}}}\n',
].join('\r\n');

const chunk1error = [
    '',
    '---',
    'Content-Type: application/json',
    'Content-Length: 104',
    '',
    '{"data":{"viewer":{"currencies":null,"user":{"profile":null}}},"errors":[{"message":"Very Bad Error"}]}\n',
].join('\r\n');

const chunk2 = [
    '',
    '---',
    'Content-Type: application/json',
    'Content-Length: 85',
    '',
    '{"path":["viewer","currencies"],"data":["USD","GBP","EUR","CAD","AUD","CHF","ðŸ˜‚"]}\n', // test unicode
].join('\r\n');

const chunk2error = [
    '',
    '---',
    'Content-Type: application/json',
    'Content-Length: 127',
    '',
    '{"path":["viewer","currencies"],"data":["USD","GBP","EUR","CAD","AUD","CHF","ðŸ˜‚"],"errors":[{"message":"Not So Bad Error"}]}\n',
].join('\r\n');

const chunk3 = [
    '',
    '---',
    'Content-Type: application/json',
    'Content-Length: 76',
    '',
    '{"path":["viewer","user","profile"],"data":{"displayName":"Steven Seagal"}}\n',
].join('\r\n');

const chunk4 = [
    '',
    '---',
    'Content-Type: application/json',
    'Content-Length: 78',
    '',
    '{"data":false,"path":["viewer","user","items","edges",1,"node","isFavorite"]}\n',
    '',
    '-----\r\n',
].join('\r\n');

describe('PathResolver', function() {
    it('should work on each chunk', function() {
        const onResponse = jest.fn();
        const resolver = new PatchResolver({
            onResponse,
        });

        resolver.handleChunk(chunk1);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();

        onResponse.mockClear();
        resolver.handleChunk(chunk2);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();

        onResponse.mockClear();
        resolver.handleChunk(chunk3);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();

        onResponse.mockClear();
        resolver.handleChunk(chunk4);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should work when chunks are split', function() {
        const onResponse = jest.fn();
        const resolver = new PatchResolver({
            onResponse,
        });

        const chunk1a = chunk1.substr(0, 35);
        const chunk1b = chunk1.substr(35, 80);
        const chunk1c = chunk1.substr(35 + 80);

        resolver.handleChunk(chunk1a);
        expect(onResponse).not.toHaveBeenCalled();
        resolver.handleChunk(chunk1b);
        expect(onResponse).not.toHaveBeenCalled();
        resolver.handleChunk(chunk1c);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
        onResponse.mockClear();

        const chunk2a = chunk2.substr(0, 35);
        const chunk2b = chunk2.substr(35);

        resolver.handleChunk(chunk2a);
        expect(onResponse).not.toHaveBeenCalled();
        resolver.handleChunk(chunk2b);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
        onResponse.mockClear();

        const chunk3a = chunk3.substr(0, 10);
        const chunk3b = chunk3.substr(11, 20);
        const chunk3c = chunk3.substr(11 + 20);

        resolver.handleChunk(chunk3a);
        expect(onResponse).not.toHaveBeenCalled();
        resolver.handleChunk(chunk3b);
        expect(onResponse).not.toHaveBeenCalled();
        resolver.handleChunk(chunk3c);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should work when chunks are combined', function() {
        const onResponse = jest.fn();
        const resolver = new PatchResolver({
            onResponse,
        });

        resolver.handleChunk(chunk1 + chunk2);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should work when chunks are combined and split', function() {
        const onResponse = jest.fn();
        const resolver = new PatchResolver({
            onResponse,
        });

        const chunk3a = chunk3.substr(0, 10);
        const chunk3b = chunk3.substr(11, 20);
        const chunk3c = chunk3.substr(11 + 20);

        resolver.handleChunk(chunk1 + chunk2 + chunk3a);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
        onResponse.mockClear();

        resolver.handleChunk(chunk3b);
        expect(onResponse).not.toHaveBeenCalled();
        resolver.handleChunk(chunk3c);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should work when chunks are combined across boundaries', function() {
        const onResponse = jest.fn();
        const resolver = new PatchResolver({
            onResponse,
        });

        const chunk2a = chunk2.substring(0, 35);
        const chunk2b = chunk2.substring(35);

        resolver.handleChunk(chunk1 + chunk2a);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
        onResponse.mockClear();
        resolver.handleChunk(chunk2b);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should merge errors', function() {
        const onResponse = jest.fn();
        const resolver = new PatchResolver({
            onResponse,
        });

        resolver.handleChunk(chunk1error);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
        onResponse.mockClear();
        resolver.handleChunk(chunk2error);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
        onResponse.mockClear();
        resolver.handleChunk(chunk3);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should work when not applying to previous', function() {
        const onResponse = jest.fn();
        const resolver = new PatchResolver({
            onResponse,
            applyToPrevious: false,
        });

        const chunk2a = chunk2.substring(0, 35);
        const chunk2b = chunk2.substring(35);

        resolver.handleChunk(chunk1 + chunk2a);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
        onResponse.mockClear();
        resolver.handleChunk(chunk2b);
        expect(onResponse.mock.calls[0][0]).toMatchSnapshot();
    });
});
