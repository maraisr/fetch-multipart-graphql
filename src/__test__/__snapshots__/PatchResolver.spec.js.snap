// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathResolver should merge errors 1`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": null,
      "user": Object {
        "profile": null,
      },
    },
  },
  "errors": Array [
    Object {
      "message": "Very Bad Error",
    },
  ],
}
`;

exports[`PathResolver should merge errors 2`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "profile": null,
      },
    },
  },
  "errors": Array [
    Object {
      "message": "Very Bad Error",
    },
    Object {
      "message": "Not So Bad Error",
    },
  ],
  "extensions": undefined,
}
`;

exports[`PathResolver should merge errors 3`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "profile": Object {
          "displayName": "Steven Seagal",
        },
      },
    },
  },
  "errors": Array [
    Object {
      "message": "Very Bad Error",
    },
    Object {
      "message": "Not So Bad Error",
    },
  ],
  "extensions": undefined,
}
`;

exports[`PathResolver should work on each chunk 1`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": null,
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": null,
      },
    },
  },
}
`;

exports[`PathResolver should work on each chunk 2`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`PathResolver should work on each chunk 3`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": Object {
          "displayName": "Steven Seagal",
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`PathResolver should work on each chunk 4`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": false,
              },
            },
          ],
        },
        "profile": Object {
          "displayName": "Steven Seagal",
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`PathResolver should work when chunks are combined 1`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`PathResolver should work when chunks are combined across boundaries 1`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": null,
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": null,
      },
    },
  },
}
`;

exports[`PathResolver should work when chunks are combined across boundaries 2`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`PathResolver should work when chunks are combined and split 1`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`PathResolver should work when chunks are combined and split 2`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": Object {
          "displayName": "Steven Seagal",
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`PathResolver should work when chunks are split 1`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": null,
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": null,
      },
    },
  },
}
`;

exports[`PathResolver should work when chunks are split 2`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`PathResolver should work when chunks are split 3`] = `
Object {
  "data": Object {
    "viewer": Object {
      "currencies": Array [
        "USD",
        "GBP",
        "EUR",
        "CAD",
        "AUD",
        "CHF",
        "ðŸ˜‚",
      ],
      "user": Object {
        "items": Object {
          "edges": Array [
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
            Object {
              "node": Object {
                "isFavorite": null,
              },
            },
          ],
        },
        "profile": Object {
          "displayName": "Steven Seagal",
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
}
`;
